### ISSUE TITLE
MSC3061: Sharing room keys for past messages

### ISSUE THREAD

**Issue by dkasak** [vector-im/element-meta#39]

**Comment by BillCarsonFr** [comment]
## SCOPE UPDATE
~~There are some scalability issues with sending the keys for the full history, as well as several edge cases. We don't want to fix all of them on existing SDKs (mobile, web).~~
~~The Rust SDK implementation should be complete.~~

~~So for existing SDKS the main goal is to  `give the new joiner the immediate context on the room to know why he was invited`, so basically the recent histoty not full history.~~

^ DEPRECATED SINCE LAST UPDATE

**Comment by benparsons** [comment]
Do we think implementation of MSC3061 should include web? https://github.com/vector-im/element-web/issues/22999 is a web UI task for room history key sharing that is currently biting a customer fairly badly. If we could include this task here it would be helpful.

**Comment by richvdh** [comment]
> Do we think implementation of MSC3061 should include web? [vector-im/element-web#22999](https://github.com/vector-im/element-web/issues/22999) is a web UI task for room history key sharing that is currently biting a customer fairly badly. If we could include this task here it would be helpful.

I'm reasonably sure this is already implemented on web. AFAICT vector-im/element-web#22999 is just a bug, not a missing implementation.

EDIT 2024-06-13: this is no longer true, because the Rust crypto stack doesn't implement it; tracked at https://github.com/element-hq/element-web/issues/26867

**Comment by thoraj** [comment]
We consider this feature important. Without it using Matrix for collaboration in a business setting is cumbersome. 
E.g when: 
* Bringing new team members to a project which is discussed in a room
* Have a co-worker cover your tasks when on vacation
* ++

What needs to happen to get this MSC moving again, so it can be merged, and find its way to clients and servers?

**Comment by jeudesprits** [comment]
I agree with what has been said above. We also chose Matrix but didn’t pay much attention to this issue at the beginning, and now we really need this feature. It would be great to hear some timelines, if possible.

**Comment by thoraj** [comment]
A guesstimate for when the security disclosure will be made would be nice. Depending on the potential attack vector the risks may be mitigated or accepted. 

Without the disclosure we don't know.

**Comment by mpeter50** [comment]
I agree that this feature is very important, but @thoraj , did you intend your last comment to be sent here? If so, what kind of security disclosure do you mean? As I understand, a vulnerability was not a topic here.

**Comment by thoraj** [comment]
@mpeter50 Sorry for not providing context. 

My reference to a disclosure  is based on a comment made in  #e2e:matrix.org today.  Apparently the MSC is blocked awaiting a security disclosure and having the issue resolved.

**Comment by dkasak** [comment]
We're working on this now, so the disclosure should come in the next couple of weeks.

**Comment by thoraj** [comment]
@dkasak Any updates for this? 

We are receiving complaints about this being flaky for our customer, and we are curious to know what the issue/disclosure is about, and how we can address this?

**Comment by dkasak** [comment]
Yes, sorry, I wanted to post an update but failed to find this particular issue.

The reason this was blocked was disclosed today in https://matrix.org/blog/2024/10/security-disclosure-matrix-js-sdk-and-matrix-react-sdk/ and the associated advisories. The plan forward is outlined in [this section](https://matrix.org/blog/2024/10/security-disclosure-matrix-js-sdk-and-matrix-react-sdk/#proposed-specification-changes).

**Comment by richvdh** [comment]
I've updated the description of this issue to reflect our current plans in this area.

**Comment by thoraj** [comment]
Very happy to see this progress, and that there are very concrete plans now 👍New members getting room history in encrypted rooms is indeed an important feature. 

I know it is not in scope atm, but I'm also wondering if the proposed exfiltration of keys using a blob, in the future could be utilized to provide keys for indexing room messages (for search purposes).

**Comment by IntelligentesTierMaulApollo13** [comment]
Maybe I missed something while digging through the whole history of this topic, so I apologize right at the beginning: 

From what I've read, my understanding is that in order to access history in encrypted rooms in the future, the inviting party must verify one of the invited party's sessions (so that the keys are actually transmitted). Now, I'm thinking of the case where a new user joins the server and needs access to, say, 30 rooms, each of which has a different moderator (and only that moderator is allowed to invite new members to the room). As I understand it, in this case the new user would have to verify their session with each of these room moderators. 

Is there any shortcut or way to automate this for environments where there are many such rooms? (I know that automating this potentially lowers security as it is then easier for an attacker to gain access to a verified session...) 

Another option I thought about is if it was possible to define a "trusted introducer" (to reuse the term from GnuPG) or a WoT-approach, but I assume there is also nothing similar, right?

**Comment by richvdh** [comment]
> From what I've read, my understanding is that in order to access history in encrypted rooms in the future, the inviting party must verify one of the invited party's sessions (so that the keys are actually transmitted). 

This is incorrect. Sorry, it's not very clear from the description at the top of this issue -- the description assumes familiarity with terminology that we use internally. So, I'll try to spell it out explicitly.

The plan relies on cross-signing by user identities and a TOFU for the identities themselves.

Suppose Alice wants to invite Bob to a room. When Alice first meets Bob, her client records Bob's public cryptographic identity. Now, when Alice sends the invite to the new room, her client will send the details of the key bundle to any of Bob's devices which *Bob* has verified, **provided** Bob's cryptographic identity hasn't changed since Alice first saw Bob. We do *not* require that Alice explicitly verify Bob or any of his devices. 

This is consistent with our plans for encrypted messaging in general; see MSC4153.

**Comment by IntelligentesTierMaulApollo13** [comment]
> ... a TOFU for the identities themselves.

It seems that this is what I was missing. Thank you for your answer!

Maybe one thing which made it easier for me to get confused is the behavior of a setting I have in my Element Web client that is labeled  `Never send encrypted messages to unverified sessions in this room from this session`. This actually forces me to explicitly verify at least one of the receiving user's sessions. So, when you were talking about verification I assumed the same meaning.

**Comment by gwennlbh** [comment]
Is there a workaround for this? I'm migrating to a new matrix account, can I "share the keys with myself" somehow?

**Comment by dkasak** [comment]
Yes, you can, from Element Web or Desktop. I *think* these keys should then (after re-import on the new device) be propagated via the Secure Backup / Key Backup / whatever-we-are-calling-it-these-days to your other devices, provided you verified them or entered your Security Key / Recovery Key.

![Image](https://github.com/user-attachments/assets/6be5f168-cd78-4b14-8b63-a9dc9951160a)

**Comment by mandulaj** [comment]
Is there any update on this? We have made several (now old and thus quite filled) encrypted rooms that new users can't see history off. For now our workaround was exporting the chat and sending it out. However this is not very convenient.

My concern:
Once implemented, would this feature be backwards compatible with old chats like ours? 
* Because if yes, we will keep using this system and once implemented, everyone will get access.

* Otherwise, we will create new unecnrypted chats and switch them back to encrypted once this features works.

Its a pity because we can't disable encryption on our old rooms.

**Comment by richvdh** [comment]
https://github.com/matrix-org/matrix-rust-sdk/pull/4700 needs to land to tell clients that it is safe to share the keys for affected messages. (And note, that PR hasn't made it into a release of matrix-rust-sdk yet, so I don't think it's been picked up by any clients.)

**Comment by mpeter50** [comment]
Oh, but its been merged, which is good news!

Do I understand it correctly that a client still needs to add its own logic for key sharing, and the sdk's next version will just add the feature that tells whether its safe to share the keys?

**Comment by mandulaj** [comment]
Well in any case as a hot fix, it would be good to add a note to the UI that warms the users that the feature doesn't work in encrypted rooms.

**Comment by richvdh** [comment]
> Do I understand it correctly that a client still needs to add its own logic for key sharing, and the sdk's next version will just add the feature that tells whether its safe to share the keys?

That's correct. We are prototyping the actual sharing logic at https://github.com/element-hq/element-meta/issues/2685.

**Comment by epchris** [comment]
Any recent updates on this? I'm still running into this issue

**Comment by epchris** [comment]
Any recent updates on this? I'm still running into this issue

**Comment by c0n4w0w** [comment]
Still no update on this issue? New users joining my channels still don't have access to any previous message history :/

**Comment by richvdh** [comment]
Tracking issue is #2829
